src/_P109_ThermOLED.ino

```diff
@@ -183,741 +118,136 @@ boolean Plugin_109(uint8_t function, struct EventStruct *event, String& string)
-        break;
-      }
-      P109_display->init(); // call to local override of init function
-      P109_display->displayOn();
-
-      uint8_t OLED_contrast = Settings.TaskDevicePluginConfig[event->TaskIndex][3];
-      P109_setContrast(OLED_contrast);
-
-      if (loglevelActiveFor(LOG_LEVEL_INFO)) {
-        String logstr = F("Thermo : Btn L:");
-        logstr += CONFIG_PIN1;
-        logstr += F("R:");
-        logstr += CONFIG_PIN2;
-        logstr += F("M:");
-        logstr += CONFIG_PIN3;
-        addLogMove(LOG_LEVEL_INFO, logstr);
-      }
-
-      for (uint8_t pin = 0; pin < 3; pin++) {
-        if (validGpio(PIN(pin))) {
-          pinMode(PIN(pin), INPUT_PULLUP);
-        }
-      }
-
-      Plugin_109_prev_temp = 99;
-
-      fs::File f = tryOpenFile(F("thermo.dat"), F("r"));
-
-      if (f) {
-        f.read(reinterpret_cast<uint8_t *>(&UserVar[event->BaseVarIndex]), 16);
-        f.close();
-      }
-      Plugin_109_lastsavetime = millis();
+      initPluginTaskData(event->TaskIndex, new (std::nothrow) P109_data_struct());
+      P109_data_struct *P109_data = static_cast<P109_data_struct *>(getPluginTaskData(event->TaskIndex));
 
-      if (UserVar[event->BaseVarIndex] < 1) {
-        UserVar[event->BaseVarIndex]     = 19; // setpoint
-        UserVar[event->BaseVarIndex + 2] = 1;  // mode (X=0,A=1,M=2)
+      if (nullptr != P109_data) {
+        success = P109_data->plugin_init(event); // Start plugin
       }
 
-      // UserVar[event->BaseVarIndex + 1] = 0; // heating (0=off,1=heating in progress)
-      // UserVar[event->BaseVarIndex + 3] = 0; // timeout (manual on for minutes)
-      if (PCONFIG(4) != -1) {
-        // pinMode(Settings.TaskDevicePluginConfig[event->TaskIndex][4], OUTPUT);
-        P109_setHeatRelay(static_cast<uint8_t>(UserVar[event->BaseVarIndex + 1]));
-      }
-
-      if (loglevelActiveFor(LOG_LEVEL_INFO)) {
-        String logstr = F("Thermo : Starting status S:");
-        logstr += toString(UserVar[event->BaseVarIndex]);
-        logstr += F(", R:");
-        logstr += toString(UserVar[event->BaseVarIndex + 1]);
-        addLogMove(LOG_LEVEL_INFO, logstr);
-      }
-
-      Plugin_109_changed    = 1;
-      Plugin_109_buttons[0] = 0; Plugin_109_buttons[1] = 0; Plugin_109_buttons[2] = 0;
-
-      //      flip screen if required
-      if (PCONFIG(1) == 2) { P109_display->flipScreenVertically(); }
-
-      //      Display the device name, logo, time and wifi
-      P109_display_header();
-      P109_display_page();
-      P109_display->display();
-      Plugin_109_init = true;
-
-      success = true;
       break;
     }
 
     case PLUGIN_EXIT:
     {
-      if (P109_display) {
-        P109_display->end();
-        delete P109_display;
-        P109_display = nullptr;
+      P109_data_struct *P109_data = static_cast<P109_data_struct *>(getPluginTaskData(event->TaskIndex));
+
+      if (nullptr != P109_data) {
+        success = P109_data->plugin_exit(event); // Stop plugin
```

Comment:
Is there a good reason not to perform everything done in `plugin_exit` in the destructor of the P109_data_struct ?

Reply:
Probably not, I'll move that to the destructor.